src/core/reactor.cc

```diff
@@ -1921,6 +1924,114 @@ reactor::inotify_add_watch(int fd, std::string_view path, uint32_t flags) {
+            throw_system_error_on(r != 0, "posix_spawn_file_actions_adddup2");
+            r = ::posix_spawn_file_actions_addclose(&actions, cout_pipe[1]);
+            throw_system_error_on(r != 0, "posix_spawn_file_actions_addclose");
+            r = ::posix_spawn_file_actions_addclose(&actions, cerr_pipe[1]);
+            throw_system_error_on(r != 0, "posix_spawn_file_actions_addclose");
+
+            r = ::posix_spawnattr_init(&attr);
+            throw_system_error_on(r != 0, "posix_spawnattr_init");
+            // make sure the following signals are not ignored by the child process
+            sigset_t default_signals;
+            sigemptyset(&default_signals);
+            sigaddset(&default_signals, SIGINT);
+            sigaddset(&default_signals, SIGTERM);
+            r = ::posix_spawnattr_setsigdefault(&attr, &default_signals);
+            throw_system_error_on(r != 0, "posix_spawnattr_setsigdefault");
+            // make sure no signals are marked in the child process
+            sigset_t mask_signals;
+            sigemptyset(&mask_signals);
+            r = ::posix_spawnattr_setsigmask(&attr, &mask_signals);
+            throw_system_error_on(r != 0, "posix_spawnattr_setsigmask");
+            r = ::posix_spawnattr_setflags(&attr, POSIX_SPAWN_SETSIGDEF | POSIX_SPAWN_SETSIGMASK);
+            throw_system_error_on(r != 0, "posix_spawnattr_setflags");
+
+            return _thread_pool->submit<syscall_result<int>>([&child_pid, &pathname, &actions, &attr, &argv, &env] {
+                std::vector<const char*> argvp;
+                std::transform(argv.cbegin(), argv.cend(), std::back_inserter(argvp),
+                               [](auto& s) { return s.c_str(); });
+                argvp.push_back(nullptr);
+                std::vector<const char*> envp;
+                std::transform(env.cbegin(), env.cend(), std::back_inserter(envp),
+                               [](auto& s) { return s.c_str(); });
+                envp.push_back(nullptr);
+                return wrap_syscall<int>(::posix_spawn(&child_pid, pathname.c_str(), &actions, &attr,
+                                                       const_cast<char* const *>(argvp.data()),
+                                                       const_cast<char* const *>(envp.data())));
+            });
+        }).then([&child_pid, &cout_pipe, &cerr_pipe, &actions, &attr] (syscall_result<int> ret) {
+            ::close(cout_pipe[1]);
+            ::close(cerr_pipe[1]);
+            posix_spawn_file_actions_destroy(&actions);
+            posix_spawnattr_destroy(&attr);
+            ret.throw_if_error();
+            return make_ready_future<std::tuple<pid_t, int, int>>(child_pid, cout_pipe[0], cerr_pipe[0]);
+        });
+    });
+}
+
+future<int> reactor::waitpid(pid_t pid) {
+    return _thread_pool->submit<syscall_result<int>>([pid] {
+        return wrap_syscall<int>(syscall(SYS_pidfd_open, pid, O_NONBLOCK));
```

Comment:
I guess it's too late to call pidfd_open() already. If it fails for whatever reason, the process is left unwaited-for

Reply:
true. will fix it.